# Mapping nucleosome occupancy 
We present a method to map nucleosomes using a DNA methyltransferase. The scripts here are used to align bisulfite-converted long Pacbio reads and predict nucleosomes on these reads. For complete details on the use and execution of this method, please refer to "[Chen et al. 2022](https://www.cell.com/cell-reports/fulltext/S2211-1247(22)01068-3?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2211124722010683%3Fshowall%3Dtrue)"
Bioinformatics and curve fitting tool boxes are required for data analysis.

## Input files
### Pacbio sequencing reads

### Reference sequences

### Positions and names of motifs


## Output files
### Forward and reverse complimentary reads
### Methylation matrix
### Nucleosome occupancy matrix
### NDR length and probability
### Methylation and nucleosome prediction bheatmaps
